packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -1641,20 +1641,30 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     _lastBottomViewInset = WidgetsBinding.instance.window.viewInsets.bottom;
   }
 
+  _WhitespaceDirectionalityFormatter _whitespaceFormatter;
+
   void _formatAndSetValue(TextEditingValue value) {
+    _whitespaceFormatter ??= _WhitespaceDirectionalityFormatter(textDirection: _textDirection);
+
     // Check if the new value is the same as the current local value, or is the same
     // as the post-formatting value of the previous pass.
     final bool textChanged = _value?.text != value?.text;
-    final bool isRepeat = value?.text == _lastFormattedUnmodifiedTextEditingValue?.text;
-    if (textChanged && !isRepeat && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
-      for (final TextInputFormatter formatter in widget.inputFormatters)
-        value = formatter.formatEditUpdate(_value, value);
-      _value = value;
-      _updateRemoteEditingValueIfNeeded();
-    } else {
+    final bool isRepeat = value == _lastFormattedUnmodifiedTextEditingValue;
+    // Always pass the text through the whitespace directionality formatter to maintain
+    // expected behavior with carets on trailing whitespace.
+    if (!isRepeat) {
+      if (textChanged && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
+        for (final TextInputFormatter formatter in widget.inputFormatters) {
+          value = formatter.formatEditUpdate(_value, value);
+        }
+        value = _whitespaceFormatter.formatEditUpdate(_value, value);
+      }
       _value = value;
     }
-    if (textChanged && widget.onChanged != null)
+    _updateRemoteEditingValueIfNeeded();
+
```

Comment:
Also, add another set of tests to check that no other currently untested side effects occur when typing exclusively RTL or LTR text (i.e. _updateRemoteEditingValueIfNeeded doesn't get called a bunch of times etc)

Reply:
Some of these are being "premerged" from https://github.com/flutter/flutter/pull/52055 as I'm working on them in parallel. I expect the other patch to land first before this one.

Reply:
This behavior should be handled in my work on the other patch, and should be replicated here.